consensus/src/quorum_store/utils.rs

```diff
@@ -188,159 +215,162 @@ impl ProofQueue {
-                        .observe((current_block_timestamp - expiration_time) as f64);
-                }
-            }
-            claims::assert_some!(self.digest_proof.remove(&digest));
-            self.digest_insertion_time.remove(&digest);
-        }
-
-        let mut ret = Vec::new();
+        let mut ret = vec![];
         let mut cur_bytes = 0;
         let mut cur_txns = 0;
-        let initial_size = self.digest_queue.len();
-        let mut size = self.digest_queue.len();
+        let mut excluded_txns = 0;
         let mut full = false;
 
-        for (digest, expiration) in self
-            .digest_queue
-            .iter()
-            .filter(|(digest, _)| !excluded_proofs.contains(digest))
-        {
-            if let Some(proof) = self
-                .digest_proof
-                .get(digest)
-                .expect("Entry for unexpired digest must exist")
-            {
-                if *expiration >= current_block_timestamp {
-                    // non-committed proof that has not expired
-                    cur_bytes += proof.num_bytes();
-                    cur_txns += proof.num_txns();
-                    if cur_bytes > max_bytes || cur_txns > max_txns {
-                        // Exceeded the limit for requested bytes or number of transactions.
-                        full = true;
-                        break;
-                    }
-                    ret.push(proof.clone());
-                    if let Some(insertion_time) = self.digest_insertion_time.get(digest) {
+        let mut author_to_num_remaining = HashMap::new();
+        for (author, batches) in self.author_to_batches.iter() {
+            author_to_num_remaining.insert(*author, batches.len());
+        }
+
+        'outer: while !author_to_num_remaining.is_empty() {
+            let shuffled_peers: Vec<_> = {
+                let mut peers: Vec<_> = author_to_num_remaining.keys().cloned().collect();
+                peers.shuffle(&mut thread_rng());
+                peers
+            };
+
+            for peer in shuffled_peers {
+                let queue = self.author_to_batches.get(&peer).unwrap();
+                let num_remaining = author_to_num_remaining.remove(&peer).unwrap();
+
+                let mut num_read = 0;
+                for i in (queue.len() - num_remaining)..queue.len() {
+                    let batch = queue.get(i).unwrap();
+                    num_read += 1;
+                    if excluded_batches.contains(batch) {
+                        excluded_txns += batch.num_txns();
+                    } else if self.batch_to_proof.get(batch).unwrap().is_some() {
+                        cur_bytes += batch.num_bytes();
+                        cur_txns += batch.num_txns();
+                        if cur_bytes > max_bytes || cur_txns > max_txns {
+                            // Exceeded the limit for requested bytes or number of transactions.
+                            full = true;
```

Comment:
Note that the block is not necessarily full. It might be the case that there is room for a smaller batch.